File: backend/render-start.js
#!/usr/bin/env node

const { spawn } = require('child_process');
const fs = require('fs');
const path = require('path');

console.log('🚀 Starting KOL Tracker Backend on Render...');
console.log('Environment:', process.env.NODE_ENV);
console.log('Port:', process.env.PORT || 10000);
console.log('Working directory:', process.cwd());

// Set up environment variables for production
const env = {
  ...process.env,
  NODE_ENV: 'production',
  PORT: process.env.PORT || 10000,
  DATABASE_URL: process.env.DATABASE_URL,
  JWT_SECRET: process.env.JWT_SECRET,
  TELETHON_URL: process.env.TELETHON_URL || 'http://localhost:8000'
};

// Log important configuration (without sensitive data)
console.log('Configuration:');
console.log('- Database URL:', env.DATABASE_URL ? 'Set ✅' : 'Missing ❌');
console.log('- JWT Secret:', env.JWT_SECRET ? 'Set ✅' : 'Missing ❌');
console.log('- Telethon URL:', env.TELETHON_URL);

// Check if simple-server.js exists
const serverPath = path.join(__dirname, 'simple-server.js');
if (!fs.existsSync(serverPath)) {
  console.error('❌ simple-server.js not found at:', serverPath);
  process.exit(1);
}

console.log('✅ Starting simple-server.js...');

// Start the server
const child = spawn('node', ['simple-server.js'], {
  cwd: __dirname,
  stdio: 'inherit',
  env: env
});

child.on('error', (error) => {
  console.error('❌ Error starting backend:', error);
  process.exit(1);
});

child.on('exit', (code) => {
  console.log(`🔄 Backend exited with code ${code}`);
  process.exit(code);
});

// Graceful shutdown
process.on('SIGTERM', () => {
  console.log('📋 Received SIGTERM, shutting down gracefully...');
  child.kill('SIGTERM');
});

process.on('SIGINT', () => {
  console.log('📋 Received SIGINT, shutting down gracefully...');
  child.kill('SIGINT');
});

console.log('🎯 Backend startup script initialized'); 

File: backend/telethon_service/render-start.py
#!/usr/bin/env python3
import os
import sys
import logging
import asyncio
from main import app, scanner
import uvicorn

# Configure logging for production
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    handlers=[
        logging.StreamHandler(sys.stdout)
    ]
)
logger = logging.getLogger(__name__)

def setup_production_environment():
    """Set up production environment variables and configuration"""
    
    # Set default values for production
    os.environ.setdefault('PORT', '10000')
    os.environ.setdefault('HOST', '0.0.0.0')
    os.environ.setdefault('SESSION_NAME', 'telegram_session')
    
    # Log environment configuration (without sensitive data)
    logger.info("🚀 Starting KOL Tracker Telethon Service on Render...")
    logger.info(f"Environment: {os.getenv('NODE_ENV', 'production')}")
    logger.info(f"Port: {os.getenv('PORT')}")
    logger.info(f"Host: {os.getenv('HOST')}")
    logger.info(f"API_ID: {'Set ✅' if os.getenv('API_ID') else 'Missing ❌'}")
    logger.info(f"API_HASH: {'Set ✅' if os.getenv('API_HASH') else 'Missing ❌'}")
    logger.info(f"Database URL: {'Set ✅' if os.getenv('DATABASE_URL') else 'Missing ❌'}")
    logger.info(f"Session Name: {os.getenv('SESSION_NAME')}")

async def startup_service():
    """Initialize the service"""
    try:
        logger.info("🔄 Initializing Telethon scanner...")
        await scanner.connect()
        logger.info("✅ Telethon service initialized successfully")
    except Exception as e:
        logger.error(f"❌ Failed to initialize service: {e}")
        # Don't exit - allow service to start for user authentication
        logger.info("⚠️ Continuing startup in authentication mode")

def main():
    """Main entry point for Render deployment"""
    
    setup_production_environment()
    
    # Get configuration from environment
    port = int(os.getenv('PORT', 10000))
    host = os.getenv('HOST', '0.0.0.0')
    
    # Add startup event to initialize scanner
    @app.on_event("startup")
    async def startup():
        await startup_service()
    
    logger.info(f"🎯 Starting server on {host}:{port}")
    
    # Start the FastAPI server
    uvicorn.run(
        "main:app",
        host=host,
        port=port,
        log_level="info",
        access_log=True,
        reload=False,  # Disable reload in production
        workers=1      # Single worker for free tier
    )

if __name__ == "__main__":
    main() 